### PR TITLE
(GH-90) Support markdownlint-cli 0.19.0 log file format

### LINKED ISSUE #90: Reading issues throws exception with markdownlint-cli 0.19.0
Readin issues from a log file created by markdownlint-cli 0.19.0 throws an exception:

```
Error: System.AggregateException: One or more errors occurred. ---> System.ArgumentOutOfRangeException: startIndex cannot be larger than length of string.
Parameter name: startIndex
   at System.String.Substring(Int32 startIndex, Int32 length)
   at Cake.Issues.Markdownlint.LogFileFormat.MarkdownlintCliLogFileFormat.TryGetFile(GroupCollection values, RepositorySettings repositorySettings, String& fileName) in C:\projects\cake-issues-markdownlint\src\Cake.Issues.Markdownlint\LogFileFormat\MarkdownlintCliLogFileFormat.cs:line 75
   at Cake.Issues.Markdownlint.LogFileFormat.MarkdownlintCliLogFileFormat.<ReadIssues>d__1.MoveNext() in C:\projects\cake-issues-markdownlint\src\Cake.Issues.Markdownlint\LogFileFormat\MarkdownlintCliLogFileFormat.cs:line 38
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Cake.Issues.IssuesReader.ReadIssues(IssueCommentFormat format) in C:\projects\cake-issues\src\Cake.Issues\IssuesReader.cs:line 56
   at Submission#0.IssueHelper.ReadMarkdownlintIssues(ICakeContext context, DirectoryPath logDirectoryPath)
   at Submission#0.<<Initialize>>b__0_46(BuildData buildData)
   at Cake.Core.CakeTaskBuilderExtensions.<>c__DisplayClass24_0.<Does>b__0(ICakeContext x)
   at Cake.Core.CakeTask.<Execute>d__43.MoveNext()
```

### PR DESCRIPTION
**Pull request by christianbumann**
Fixes #90

### CODE REVIEW

File: src/Cake.Issues.Markdownlint.Tests/LogFileFormat/MarkdownlintLogfileHelperTests.cs [hunk 1]
```diff
@@ -0,0 +1,59 @@
+﻿namespace Cake.Issues.Markdownlint.Tests.LogFileFormat
+{
+    using Cake.Issues.Markdownlint.LogFileFormat;
+    using Cake.Issues.Testing;
+    using Shouldly;
+    using Xunit;
+
+    public sealed class MarkdownlintLogFileHelperTests
+    {
+        public sealed class TheCtor
```
Comment (pascalberger) [inline src/Cake.Issues.Markdownlint.Tests/LogFileFormat/MarkdownlintLogfileHelperTests.cs]: As this tests the `GetRegexForMarkdownIssueLine` the class should be name `TheGetRegexForMarkdownIssueLineMethod`

File: src/Cake.Issues.Markdownlint.Tests/LogFileFormat/MarkdownlintLogfileHelperTests.cs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        {
+            [Fact]
+            public void GetRegexForMarkdownIssueLine_Should_Return_Correct_Pattern_For_Markdownlint_Cli_Smaller_Than_0_19_0()
```
Comment (pascalberger) [inline src/Cake.Issues.Markdownlint.Tests/LogFileFormat/MarkdownlintLogfileHelperTests.cs]: Remove `GetRegexForMarkdownIssueLine_`

File: src/Cake.Issues.Markdownlint.Tests/LogFileFormat/MarkdownlintLogfileHelperTests.cs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+                // Issue line Markdownlint-cli version < 0.19.0
+                var issueLine = "C:/Temp/index.md: 5: MD025/single-title/single-h1 Multiple top level headings in the same document [Context: \"# Markdown Study Guide\"]";
+                var result = MarkdownlintLogFileHelper.GetRegexForMarkdownIssueLine(issueLine);
+
+                // Then
+                result.ToString().Equals(@"(.*): (\d *): (MD\d *)/ ((?:\w * -*/*)*) (.*)");
+            }
+
+            [Fact]
+            public void GetRegexForMarkdownIssueLineShould_Return_Correct_Pattern_For_Markdownlint_Cli_Bigger_Equals_Than_0_19_0()
```
Comment (pascalberger) [inline src/Cake.Issues.Markdownlint.Tests/LogFileFormat/MarkdownlintLogfileHelperTests.cs]: Remove `GetRegexForMarkdownIssueLine_`

File: src/Cake.Issues.Markdownlint.Tests/LogFileFormat/MarkdownlintLogfileHelperTests.cs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+                // Issue line Markdownlint-cli version >= 0.19.0
+                var issueLine = "C:/Temp/index.md:5 MD025/single-title/single-h1 Multiple top level headings in the same document [Context: \"# Markdown Study Guide\"]";
+                var result = MarkdownlintLogFileHelper.GetRegexForMarkdownIssueLine(issueLine);
+
+                // Then
+                result.ToString().Equals(@"(.*):(\d *) (MD\d *)/ ((?:\w * -*/*)*) (.*)");
+            }
+
+            [Fact]
+            public void GetRegexForMarkdownIssueLineShould_Throw_Exception_For_Wrong_Format()
```
Comment (pascalberger) [inline src/Cake.Issues.Markdownlint.Tests/LogFileFormat/MarkdownlintLogfileHelperTests.cs]: Remove `GetRegexForMarkdownIssueLine`

File: src/Cake.Issues.Markdownlint.Tests/LogFileFormat/MarkdownlintLogfileHelperTests.cs [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+    public sealed class MarkdownlintLogFileHelperTests
+    {
+        public sealed class TheCtor
+        {
+            [Fact]
+            public void GetRegexForMarkdownIssueLine_Should_Return_Correct_Pattern_For_Markdownlint_Cli_Smaller_Than_0_19_0()
+            {
+                // Given / When
+                // Issue line Markdownlint-cli version < 0.19.0
```
Comment (pascalberger) [inline src/Cake.Issues.Markdownlint.Tests/LogFileFormat/MarkdownlintLogfileHelperTests.cs]: Comment is not required. Test case name should be sufficient (Same for other test cases)

File: src/Cake.Issues.Markdownlint.Tests/LogFileFormat/MarkdownlintLogfileHelperTests.cs [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    using Xunit;
+
+    public sealed class MarkdownlintLogFileHelperTests
+    {
+        public sealed class TheCtor
+        {
+            [Fact]
+            public void GetRegexForMarkdownIssueLine_Should_Return_Correct_Pattern_For_Markdownlint_Cli_Smaller_Than_0_19_0()
+            {
+                // Given / When
```
Comment (pascalberger) [inline src/Cake.Issues.Markdownlint.Tests/LogFileFormat/MarkdownlintLogfileHelperTests.cs]: Add `// Given` comment for the `var issueLine = ...` and `// When` comment for the `var result =...` line. (Same for other test cases)

File: src/Cake.Issues.Markdownlint.Tests/LogFileFormat/MarkdownlintLogfileHelperTests.cs [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+                // Unsupported issue line
+                var issueLine = "UnsupportedFormat";
+                var result = Record.Exception(() => MarkdownlintLogFileHelper.GetRegexForMarkdownIssueLine(issueLine));
+
+                // Then
+                result.Message.ShouldBe("No RegEx found for the Markdown issue line.");
+            }
+
+            [Fact]
+            public void GetRegexForMarkdownIssueLineShould_Throw_Exception_For_Nothing_Parameter()
```
Comment (pascalberger) [inline src/Cake.Issues.Markdownlint.Tests/LogFileFormat/MarkdownlintLogfileHelperTests.cs]: Remove `GetRegexForMarkdownIssueLine`

File: src/Cake.Issues.Markdownlint/LogFileFormat/MarkdownlintLogfileHelper.cs [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    /// Helper class for Markdownlint log files.
+    /// </summary>
+    internal static class MarkdownlintLogFileHelper
+    {
+        /// <summary>
+        /// Reads the RegEx for a Mardown issue line depending its line format.
+        /// </summary>
+        /// <param name="markdownIssueLine">Markdown issue line.</param>
+        /// <returns>The RegEx pattern to parse the Markdown issue line.</returns>
+        public static Regex GetRegexForMarkdownIssueLine(string markdownIssueLine)
```
Comment (pascalberger) [inline src/Cake.Issues.Markdownlint/LogFileFormat/MarkdownlintLogfileHelper.cs]: We don't need a static helper class for this. You can move this as a private method to the `MarkdownlintCliLogFileFormat` class.

File: src/Cake.Issues.Markdownlint/LogFileFormat/MarkdownlintLogfileHelper.cs [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+                regex = new Regex(@"(.*): (\d*): (MD\d*)/((?:\w*-*/*)*) (.*)");
+                regexMatch = regex.Match(markdownIssueLine);
+            }
+
+            if (regexMatch.Success)
+            {
+                return regex;
+            }
+
+            throw new ArgumentException("No RegEx found for the Markdown issue line.");
```
Comment (pascalberger) [inline src/Cake.Issues.Markdownlint/LogFileFormat/MarkdownlintLogfileHelper.cs]: This should not be an `ArgumentException`. Use `InvalidOperationException` (or a custom exception)

**Comment by christianbumann** [comment]
@pascalberger Changes done, I had to improve the Regex to working more reliant. Strange on server all tests are green, but on my local machine one test is faling. Can you try locally?

**Comment by pascalberger** [comment]
@christianbumann Test cases work fine on my side. What is the error message you get?

**Comment by christianbumann** [comment]
`A total of 1 test files matched the specified pattern.
[xUnit.net 00:00:01.46]     Cake.Issues.Markdownlint.Tests.MarkdownlintRuleUrlResolverTests+TheResolveRuleUrlMethod.Should_Parse_RuleId(rule: "MD0001", expectedRuleId: 1) [FAIL]
  X Cake.Issues.Markdownlint.Tests.MarkdownlintRuleUrlResolverTests+TheResolveRuleUrlMethod.Should_Parse_RuleId(rule: "MD0001", expectedRuleId: 1) [113ms]
  Error Message:
   Shouldly.ShouldAssertException : Shouldly uses your source code to generate its great error messages, build your test project with full debug information to get better error messages
The provided expression`

**Comment by pascalberger** [comment]
@christianbumann There's actually no reason to use different regular expressions. Parsing can be done with a single regular expression. I've updated the code accordingly.

**Comment by pascalberger** [comment]
@christianbumann The flaky test on local test execution seems to be an issue with parallel test execution. I'll tackle this on a separate PR

**Pull request merged**